highfidelity/hifi
# Touch feeling for near grab

This feature is an attempt to improve the user experience when interacting with entities using the avatar's hands.  To accomplish this, the hand's pose adapts to the entity's mesh that is within a certain touch-distance.


# Test Plan
(note: it's expected that hands and fingers penetrate the surface sometimes when the imaginary plane created by the hand palm intersect the object)
- Make sure your avatar has five fingers on each hand (the wood mannequin should also work but it's not appropriate for testing)
- Load the script: system/controllers/handTouch.js
- Go to a domain where you can edit and import a complex model : https://raw.githubusercontent.com/KhronosGroup/glTF-Sample-Models/master/2.0/DamagedHelmet/glTF/DamagedHelmet.gltf
- Grab it with the left hand and bring it near you so you can start touching it with the right hand.
- When your hand go near the object the fingers should adapt to the surface. Move your hand around the object's surface. The fingers should adapt to the surface accordingly. 
- Click the trigger meanwhile your fingers are touching the object. Your fingers should close slightly over the object giving the impression that you are grabbing it.
- Repeat the process this time touching the object with the left hand. The outcome should be similar.
- After scaling the object everything should work as expected. (objects smaller than a ping pong ball would be difficult to intersect )
- After scaling your avatar everything should work as expected.
- After changing avatars everything should work as expected.
- Apps that override hand poses should still work: Get the Fingerpaint app from the market. When using the app make sure that your finger is painting with the index finger and the hand pose is as expected.
- Turn the script off meanwhile touching something. After that the hands default poses should work as expected.

Also test this:
https://highfidelity.testrail.net/index.php?/suites/view/20&group_by=cases:section_id&group_order=asc

In addition, please run this test plan with the script turned off - https://highfidelity.testrail.net/index.php?/plans/view/1893

# Known Issues

- It needs a better integration with ControllerDispatcher.js
